FantasmoSDK/Classes/Models/Analytics/TotalDeviceTranslationAccumulator.swift

```diff
@@ -0,0 +1,74 @@
+//
+//  ARKitDeviceMotion.swift
+//  FantasmoSDK
+//
+//  Created by Yauheni Klishevich on 14.07.2021.
+//
+
+import ARKit
+
+/// Calculates the trajectory length of a device based on the device transform provided by the sequence of ARKit frames.
+///
+/// To reduce error introduced by noise we use decimation (downsampling without passing through low-pass filter) by an integer factor as the most
+/// simple approach which should yield sufficient accuracy.
+/// More info about downsampling and decimation can be read at https://en.wikipedia.org/wiki/Downsampling_(signal_processing)
+public class TotalDeviceTranslationAccumulator {
+    
+    /// Current value of total translation in meters, which is updated as more frames are passed via `update(`
+    private(set) var totalTranslation: Float = 0.0 {
+        didSet {
+            totalTranslationObservers.forEach { _, closure in
+                closure(totalTranslation)
+            }
+        }
+    }
+    
+    /// Number of each frame in a frame sequence which should be taken for calculating total device translation.
+    private(set) var decimationFactor: UInt = 10
+    
+    private var frameCounter: UInt = 0
+    private var nextFrameToTake: UInt = 0
+    private var previousTranslation: simd_float3!
+    private var totalTranslationObservers = [UUID : ((Float) -> Void)]()
+    
+    public init(decimationFactor: UInt = 10) {
+        self.decimationFactor = decimationFactor
+    }
+
+    func update(with nextFrame: ARFrame) {
+        if previousTranslation == nil {
+            previousTranslation = nextFrame.camera.transform.translation
+        }
+        
+        if frameCounter == nextFrameToTake {
```

Comment:
Maybe this should be `>=` in case one day this update can be occasionally disabled and frame numbers don't line up exactly?

Reply:
I think that > sign will be confusing here. The reader will be trying to understand why and in which scenario `>` will work out. And after some additional investigation will be forced to deduce that it was added only to be on the safe side.

Also cannot understand what disabling can destroy lining up.
Could you clarify your point?

Reply:
I explained it badly. A common error in situations like this is when `frameCounter` is accidentally increased so as to exceed `nextFrameToTake` and the condition never becomes true, and updates stop. That clearly will not happen in the code as you have it now, but it's the kind of bug that can creep in. I often use a defensive `>=` just to be safe.

Reply:
Ok, I will add > and will add comment with explicitation - https://take.ms/sjWkq